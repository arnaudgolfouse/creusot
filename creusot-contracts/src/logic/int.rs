use crate::{
    ghost::Plain,
    logic::ops::{AddLogic, DivLogic, MulLogic, NegLogic, RemLogic, SubLogic},
    std::ops::{Add, Div, Mul, Neg, Rem, Sub},
    *,
};

/// An unbounded, mathematical integer.
///
/// This type cannot be only be constructed in logical or ghost code.
///
/// # Integers in pearlite
///
/// Note that in pearlite, all integer literals are of type `Int`:
/// ```
/// # use creusot_contracts::*;
/// let x = 1i32;
/// //             â†“ need to use the view operator to convert `i32` to `Int`
/// proof_assert!(x@ == 1);
/// ```
///
/// You can use the usual operators on integers: `+`, `-`, `*`, `/` and `%`.
///
/// Note that those operators are _not_ available in ghost code.
#[intrinsic("int")]
#[builtin("int")]
pub struct Int;

impl Clone for Int {
    #[check(ghost)]
    #[ensures(result == *self)]
    fn clone(&self) -> Self {
        *self
    }
}
impl Copy for Int {}
#[trusted]
impl Plain for Int {}

// Logical functions
impl Int {
    /// Compute `self^p`.
    ///
    /// # Example
    ///
    /// ```
    /// # use creusot_contracts::*;
    /// proof_assert!(3.pow(4) == 729);
    /// ```
    #[logic]
    #[builtin("int.Power.power")]
    #[allow(unused_variables)]
    pub fn pow(self, p: Int) -> Int {
        dead
    }

    /// Compute `2^p`.
    ///
    /// # Example
    ///
    /// ```
    /// # use creusot_contracts::*;
    /// proof_assert!(pow2(4) == 16);
    /// ```
    #[logic]
    #[builtin("bv.Pow2int.pow2")]
    #[allow(unused_variables)]
    pub fn pow2(self) -> Int {
        dead
    }

    /// Compute the maximum of `self` and `x`.
    ///
    /// # Example
    ///
    /// ```
    /// # use creusot_contracts::*;
    /// proof_assert!(10.max(2) == 10);
    /// ```
    #[logic]
    #[builtin("int.MinMax.max")]
    #[allow(unused_variables)]
    pub fn max(self, x: Int) -> Int {
        dead
    }

    /// Compute the minimum of `self` and `x`.
    ///
    /// # Example
    ///
    /// ```
    /// # use creusot_contracts::*;
    /// proof_assert!(10.max(2) == 2);
    /// ```
    #[logic]
    #[builtin("int.MinMax.min")]
    #[allow(unused_variables)]
    pub fn min(self, x: Int) -> Int {
        dead
    }

    /// Compute the euclidean division of `self` by `d`.
    ///
    /// # Example
    ///
    /// ```
    /// # use creusot_contracts::*;
    /// proof_assert!(10.div_euclid(3) == 3);
    /// ```
    #[logic]
    #[builtin("int.EuclideanDivision.div")]
    #[allow(unused_variables)]
    pub fn div_euclid(self, d: Int) -> Int {
        dead
    }

    /// Compute the remainder of the euclidean division of `self` by `d`.
    ///
    /// # Example
    ///
    /// ```
    /// # use creusot_contracts::*;
    ///  proof_assert!(10.rem_euclid(3) == 1);
    /// ```
    #[logic]
    #[builtin("int.EuclideanDivision.mod")]
    #[allow(unused_variables)]
    pub fn rem_euclid(self, d: Int) -> Int {
        dead
    }

    /// Compute the absolute difference of `self` and `x`.
    ///
    /// # Example
    ///
    /// ```
    /// # use creusot_contracts::*;
    /// proof_assert!(10.abs_diff(3) == 7);
    /// proof_assert!(3.abs_diff(10) == 7);
    /// proof_assert!((-5).abs_diff(5) == 10);
    /// ```
    #[logic(open)]
    pub fn abs_diff(self, other: Int) -> Int {
        if self < other { other - self } else { self - other }
    }
}

impl AddLogic for Int {
    type Output = Self;
    #[logic]
    #[builtin("mach.int.Int.(+)")]
    #[allow(unused_variables)]
    fn add(self, other: Self) -> Self {
        dead
    }
}

impl SubLogic for Int {
    type Output = Self;
    #[logic]
    #[builtin("mach.int.Int.(-)")]
    #[allow(unused_variables)]
    fn sub(self, other: Self) -> Self {
        dead
    }
}

impl MulLogic for Int {
    type Output = Self;
    #[logic]
    #[builtin("mach.int.Int.(*)")]
    #[allow(unused_variables)]
    fn mul(self, other: Self) -> Self {
        dead
    }
}

impl DivLogic for Int {
    type Output = Self;
    #[logic]
    #[builtin("mach.int.Int.div")]
    #[allow(unused_variables)]
    fn div(self, other: Self) -> Self {
        dead
    }
}

impl RemLogic for Int {
    type Output = Self;
    #[logic]
    #[builtin("mach.int.Int.mod")]
    #[allow(unused_variables)]
    fn rem(self, other: Self) -> Self {
        dead
    }
}

impl NegLogic for Int {
    type Output = Self;
    #[logic]
    #[builtin("mach.int.Int.(-_)")]
    fn neg(self) -> Self {
        dead
    }
}

// ========== Ghost operations =============

// Ghost functions
impl Int {
    /// Create a new `Int` value
    ///
    /// The result is wrapped in a [`Ghost`], so that it can only be access inside a
    /// [`ghost!`] block.
    ///
    /// You should not have to use this method directly: instead, use the `int` suffix
    /// inside of a `ghost` block:
    /// ```
    /// # use creusot_contracts::*;
    /// let x: Ghost<Int> = ghost!(1int);
    /// ghost! {
    ///     let y: Int = 2int;
    /// };
    /// ```
    #[trusted]
    #[check(ghost)]
    #[ensures(*result == value@)]
    #[allow(unreachable_code)]
    #[allow(unused_variables)]
    pub fn new(value: i128) -> Ghost<Self> {
        Ghost::conjure()
    }

    #[trusted]
    #[check(ghost)]
    #[ensures(^self == *self + 1)]
    pub fn incr_ghost(&mut self) {}

    #[trusted]
    #[check(ghost)]
    #[ensures(^self == *self - 1)]
    pub fn decr_ghost(&mut self) {}
}

impl PartialEq for Int {
    #[trusted]
    #[check(ghost)]
    #[ensures(result == (*self == *other))]
    #[allow(unused_variables)]
    fn eq(&self, other: &Self) -> bool {
        panic!()
    }
}

impl PartialOrd for Int {
    #[trusted]
    #[check(ghost)]
    #[ensures(result == Some((*self).cmp_log(*other)))]
    #[allow(unused_variables)]
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        panic!()
    }

    #[trusted]
    #[check(ghost)]
    #[ensures(result == (*self).lt_log(*other))]
    #[allow(unused_variables)]
    fn lt(&self, other: &Self) -> bool {
        panic!()
    }

    #[trusted]
    #[check(ghost)]
    #[ensures(result == (*self).le_log(*other))]
    #[allow(unused_variables)]
    fn le(&self, other: &Self) -> bool {
        panic!()
    }

    #[trusted]
    #[check(ghost)]
    #[ensures(result == (*self).gt_log(*other))]
    #[allow(unused_variables)]
    fn gt(&self, other: &Self) -> bool {
        panic!()
    }

    #[trusted]
    #[check(ghost)]
    #[ensures(result == (*self).ge_log(*other))]
    #[allow(unused_variables)]
    fn ge(&self, other: &Self) -> bool {
        panic!()
    }
}

impl Add for Int {
    type Output = Int;
    #[trusted]
    #[check(ghost)]
    #[ensures(result == self + other)]
    #[allow(unused_variables)]
    fn add(self, other: Int) -> Self {
        panic!()
    }
}

impl Sub for Int {
    type Output = Int;
    #[trusted]
    #[check(ghost)]
    #[ensures(result == self - other)]
    #[allow(unused_variables)]
    fn sub(self, other: Int) -> Self {
        panic!()
    }
}

impl Mul for Int {
    type Output = Int;
    #[trusted]
    #[check(ghost)]
    #[ensures(result == self * other)]
    #[allow(unused_variables)]
    fn mul(self, other: Int) -> Self {
        panic!()
    }
}

impl Div for Int {
    type Output = Int;
    #[trusted]
    #[check(ghost)]
    #[ensures(result == self / other)]
    #[allow(unused_variables)]
    fn div(self, other: Int) -> Self {
        panic!()
    }
}

impl Rem for Int {
    type Output = Int;
    #[trusted]
    #[check(ghost)]
    #[ensures(result == self % other)]
    #[allow(unused_variables)]
    fn rem(self, other: Int) -> Self {
        panic!()
    }
}

impl Neg for Int {
    type Output = Int;
    #[trusted]
    #[check(ghost)]
    #[ensures(result == -self)]
    fn neg(self) -> Self {
        panic!()
    }
}
