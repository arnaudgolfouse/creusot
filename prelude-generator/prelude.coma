module Any
  function any_l (a :'a) : 'b
end

module MutBorrow
  type t 'a = { current : 'a; final : 'a; id : int }

  let borrow_final < 'a > (a : 'a) (id : int) (ret (result : t 'a)) =
    any [ k (fin : 'a) -> (! ret { { current = a; final = fin; id = id } })]

  let borrow_mut < 'a > (a : 'a) (ret (result : t 'a)) =
    any [ k (fin : 'a) (id: int) -> (! ret { { current = a; final = fin; id = id } })]

  function borrow_logic (cur fin : 'a) (id : int) : t 'a = { current = cur; final = fin; id = id }

  function get_id (x : t 'a) : int = x.id
  function inherit_id (old_id inherit_path: int) : int
end

module Bool
  use bool.Bool

  function to_int [@inline:trivial] (b : bool) : int = if b then 1 else 0

  function bw_and [@inline:trivial] (a:bool) (b:bool) : bool = a /\ b
  function bw_or [@inline:trivial] (a:bool) (b:bool) : bool = a \/ b
  function bw_xor [@inline:trivial] (a:bool) (b:bool) : bool = not a = b

  predicate le [@inline:trivial] (a:bool) (b:bool) = (not a) \/ b
  predicate lt [@inline:trivial] (a:bool) (b:bool) = (not a) /\ b
  predicate ge [@inline:trivial] (a:bool) (b:bool) = a \/ not b
  predicate gt [@inline:trivial] (a:bool) (b:bool) = a /\ not b
end

module Char
  (* utf8 characters (not glyphs)
      highly restricted until Why3 supports Unicode strings
  *)
  use int.Int
  use bv.BV256 as BV256

  type t

  function to_int (x : t) : int
  axiom to_int: forall c. 0 <= to_int c < 0x10FFFF /\ not(0xD800 <= to_int c < 0xE000)
  axiom extensionality:
    forall x y: t [to_int x, to_int y]. to_int x = to_int y -> x = y

  function of_int (n: int) : t
  axiom to_of_int: forall n [of_int n]. 0 <= n < 0x10FFFF /\ not(0xD800 <= n < 0xE000) -> to_int (of_int n) = n

  predicate le [@inline:trivial] (a:t) (b:t) = to_int a <= to_int b
  predicate lt [@inline:trivial] (a:t) (b:t) = to_int a < to_int b
  predicate ge [@inline:trivial] (a:t) (b:t) = to_int a >= to_int b
  predicate gt [@inline:trivial] (a:t) (b:t) = to_int a > to_int b

  let of_int (n : int) { 0 <= n < 0x10FFFF } { not(0xD800 <= n < 0xE000) }
    (ret (result : t) { to_int result = n }) = any

  let of_BV256 (n: BV256.t)
    { [@expl:arithmetic overflow] BV256.t'int n >= 0 /\ BV256.t'int n < 0x10FFFF }
    { not(0xD800 <= BV256.t'int n < 0xE000) }
    (ret (result: t)  { to_int result = BV256.t'int n}) = any
end

module Opaque
  type ptr
  let fresh_ptr (ret (result : ptr) {true}) = any

  type dyn
  type foreign
end
