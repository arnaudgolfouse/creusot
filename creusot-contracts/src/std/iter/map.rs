use crate::prelude::*;
#[cfg(creusot)]
use crate::resolve::structural_resolve;
use std::iter::Map;

pub trait MapExt<I, F> {
    #[logic]
    fn iter(self) -> I;

    #[logic]
    fn func(self) -> F;
}

impl<I, F> MapExt<I, F> for Map<I, F> {
    #[logic(opaque)]
    fn iter(self) -> I {
        dead
    }

    #[logic(opaque)]
    fn func(self) -> F {
        dead
    }
}

impl<I, F> Resolve for Map<I, F> {
    #[logic(open, prophetic, inline)]
    fn resolve(self) -> bool {
        resolve(self.iter()) && resolve(self.func())
    }

    #[trusted]
    #[logic(prophetic)]
    #[requires(structural_resolve(self))]
    #[ensures(self.resolve())]
    fn resolve_coherence(self) {}
}

impl<I: IteratorSpec, B, F: FnMut(I::Item) -> B> Invariant for Map<I, F> {
    #[logic(prophetic)]
    #[ensures(result ==> inv(self.iter()) && inv(self.func()))]
    fn invariant(self) -> bool {
        pearlite! {
            inv(self.iter()) && inv(self.func()) &&
            reinitialize::<I, B, F>() &&
            preservation(self.iter(), self.func()) &&
            next_precondition(self.iter(), self.func())
        }
    }
}

impl<I: IteratorSpec, B, F: FnMut(I::Item) -> B> IteratorSpec for Map<I, F> {
    #[logic(open, prophetic)]
    fn completed(&mut self) -> bool {
        pearlite! {
            (exists<inner: &mut _> *inner == self.iter() && ^inner == (^self).iter() && inner.completed())
            && (*self).func() == (^self).func()
        }
    }

    #[logic(open, prophetic, inline)]
    fn produces(self, visited: Seq<Self::Item>, succ: Self) -> bool {
        pearlite! {
            self.func().hist_inv(succ.func())
            && exists<fs: Seq<&mut F>> fs.len() == visited.len()
            && exists<s: Seq<I::Item>>
                #[trigger(self.iter().produces(s, succ.iter()))]
                s.len() == visited.len() && self.iter().produces(s, succ.iter())
            && (forall<i> 1 <= i && i < fs.len() ==>  ^fs[i - 1] == *fs[i])
            && if visited.len() == 0 { self.func() == succ.func() }
               else { *fs[0] == self.func() &&  ^fs[visited.len() - 1] == succ.func() }
            && forall<i> 0 <= i && i < visited.len() ==>
                 self.func().hist_inv(*fs[i])
                 && (*fs[i]).precondition((s[i],))
                 && (*fs[i]).postcondition_mut((s[i],), ^fs[i], visited[i])
        }
    }

    #[logic(law)]
    #[ensures(self.produces(Seq::empty(), self))]
    fn produces_refl(self) {}

    #[logic(law)]
    #[requires(a.produces(ab, b))]
    #[requires(b.produces(bc, c))]
    #[ensures(a.produces(ab.concat(bc), c))]
    fn produces_trans(a: Self, ab: Seq<Self::Item>, b: Self, bc: Seq<Self::Item>, c: Self) {}
}

#[logic(open, prophetic)]
pub fn next_precondition<I: IteratorSpec, B, F: FnMut(I::Item) -> B>(iter: I, func: F) -> bool {
    pearlite! {
        forall<e: I::Item, i: I>
            #[trigger(iter.produces(Seq::singleton(e), i))]
            iter.produces(Seq::singleton(e), i) ==>
            func.precondition((e,))
    }
}

#[logic(open, prophetic)]
pub fn preservation<I: IteratorSpec, B, F: FnMut(I::Item) -> B>(iter: I, func: F) -> bool {
    pearlite! {
        forall<s: Seq<I::Item>, e1: I::Item, e2: I::Item, f: &mut F, b: B, i: I>
            #[trigger(iter.produces(s.push_back(e1).push_back(e2), i), (*f).postcondition_mut((e1,), ^f, b))]
            func.hist_inv(*f) ==>
            iter.produces(s.push_back(e1).push_back(e2), i) ==>
            (*f).precondition((e1,)) ==>
            (*f).postcondition_mut((e1,), ^f, b) ==>
            (^f).precondition((e2, ))
    }
}

#[logic(open, prophetic)]
pub fn reinitialize<I: IteratorSpec, B, F: FnMut(I::Item) -> B>() -> bool {
    pearlite! {
        forall<iter: &mut I, func: F>
            iter.completed() ==> next_precondition(^iter, func) && preservation(^iter, func)
    }
}
