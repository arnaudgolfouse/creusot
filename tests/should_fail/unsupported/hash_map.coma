module M_hash_map__qyi5744717520881225036__new [#"hash_map.rs" 18 4 18 24] (* Foo *)
  let%span shash_map = "hash_map.rs" 19 13 19 27
  
  use creusot.int.UInt64
  use creusot.prelude.Opaque
  use creusot.prelude.Any
  
  type t_RandomState = { t_RandomState__k0: UInt64.t; t_RandomState__k1: UInt64.t }
  
  type t_NonNull = { t_NonNull__pointer: Opaque.ptr }
  
  type t_RawTableInner = {
    t_RawTableInner__bucket_mask: UInt64.t;
    t_RawTableInner__ctrl: t_NonNull;
    t_RawTableInner__growth_left: UInt64.t;
    t_RawTableInner__items: UInt64.t }
  
  type t_RawTable = { t_RawTable__table: t_RawTableInner; t_RawTable__alloc: (); t_RawTable__marker: () }
  
  type t_HashMap = { t_HashMap__hash_builder: t_RandomState; t_HashMap__table: t_RawTable }
  
  type t_HashMap'0 = { t_HashMap__base: t_HashMap }
  
  type t_Foo = { t_Foo__0: t_HashMap'0 }
  
  meta "compute_max_steps" 1000000
  
  meta "select_lsinst" "all"
  
  let rec new [#"hash_map.rs" 18 4 18 24] (return' (x: t_Foo)) = (! bb0
    [ bb0 = {[%#shash_map] false} any ] [ & _0: t_Foo = Any.any_l () | & _1: t_HashMap'0 = Any.any_l () ])
    [ return''0 (result: t_Foo) -> (! return' {result}) ]
end
module M_hash_map__qyi5744717520881225036__add [#"hash_map.rs" 23 4 23 44] (* Foo *)
  let%span shash_map = "hash_map.rs" 24 9 24 26
  let%span shash_map'0 = "hash_map.rs" 22 14 22 45
  let%span shash_map'1 = "hash_map.rs" 11 4 11 25
  let%span sfmap = "../../../creusot-contracts/src/logic/fmap.rs" 92 4 92 26
  
  use creusot.prelude.Any
  use creusot.prelude.MutBorrow
  use creusot.int.UInt64
  use creusot.prelude.Opaque
  use creusot.int.UInt8
  use map.Map
  
  type t_RandomState = { t_RandomState__k0: UInt64.t; t_RandomState__k1: UInt64.t }
  
  type t_NonNull = { t_NonNull__pointer: Opaque.ptr }
  
  type t_RawTableInner = {
    t_RawTableInner__bucket_mask: UInt64.t;
    t_RawTableInner__ctrl: t_NonNull;
    t_RawTableInner__growth_left: UInt64.t;
    t_RawTableInner__items: UInt64.t }
  
  type t_RawTable = { t_RawTable__table: t_RawTableInner; t_RawTable__alloc: (); t_RawTable__marker: () }
  
  type t_HashMap = { t_HashMap__hash_builder: t_RandomState; t_HashMap__table: t_RawTable }
  
  type t_HashMap'0 = { t_HashMap__base: t_HashMap }
  
  type t_Foo = { t_Foo__0: t_HashMap'0 }
  
  type t_FMap
  
  type t_Option = C_None | C_Some UInt8.t
  
  function view (self: t_FMap) : Map.map int t_Option
  
  function get [@inline:trivial] (self: t_FMap) (k: int) : t_Option = [%#sfmap] Map.get (view self) k
  
  meta "rewrite_def" function get
  
  function view'0 (self: t_HashMap'0) : t_FMap
  
  function view'1 [#"hash_map.rs" 12 4 12 33] (self: t_Foo) : t_FMap = [%#shash_map'1] view'0 self.t_Foo__0
  
  meta "compute_max_steps" 1000000
  
  meta "select_lsinst" "all"
  
  let rec add [#"hash_map.rs" 23 4 23 44] (self: MutBorrow.t t_Foo) (num: UInt64.t) (bar: UInt8.t) (return' (x: ())) =
    (! bb0
    [ bb0 = {[%#shash_map] false} any ]
    [ & _0: () = Any.any_l ()
    | & self'0: MutBorrow.t t_Foo = self
    | & num'0: UInt64.t = num
    | & bar'0: UInt8.t = bar
    | & _6: MutBorrow.t UInt8.t = Any.any_l () ])
    [ return''0 (result: ()) -> {[@expl:add ensures] [%#shash_map'0] get (view'1 self.final) (UInt64.t'int num)
      = C_Some bar}
      (! return' {result}) ]
end
