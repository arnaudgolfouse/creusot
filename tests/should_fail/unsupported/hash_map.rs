// WHY3PROVE
extern crate creusot_contracts;
use ::std::collections::HashMap;
use creusot_contracts::*;

pub struct Foo(HashMap<u64, u8>);

impl View for Foo {
    type ViewTy = <HashMap<u64, u8> as View>::ViewTy; //FMap<Int, u8>;

    #[logic(open(crate), inline)]
    fn view(self) -> Self::ViewTy {
        self.0.view()
    }
}

impl Foo {
    pub fn new() -> Self {
        Self(HashMap::new())
    }

    #[ensures((^self)@.get(num@) == Some(bar))]
    pub fn add(&mut self, num: u64, bar: u8) {
        *self.0.entry(num).or_insert(bar) = bar;
    }
}
