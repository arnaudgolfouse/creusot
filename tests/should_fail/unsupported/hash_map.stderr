warning: calling external function `new` with no contract will yield an impossible precondition
  --> hash_map.rs:19:14
   |
19 |         Self(HashMap::new())
   |              ^^^^^^^^^^^^^^ function called here
   |
   = note: `#[warn(creusot::contractless_external_function)]` on by default

warning: calling external function `entry` with no contract will yield an impossible precondition
  --> hash_map.rs:24:10
   |
24 |         *self.0.entry(num).or_insert(bar) = bar;
   |          ^^^^^^^^^^^^^^^^^ function called here

warning: calling external function `or_insert` with no contract will yield an impossible precondition
  --> hash_map.rs:24:10
   |
24 |         *self.0.entry(num).or_insert(bar) = bar;
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ function called here

warning: 3 warnings emitted

